From 737b49208bca2d382aa4a0707ff288cf8cb49da9 Mon Sep 17 00:00:00 2001
From: Antony Abee Prakash XV <antonyabee.prakashxv@adlinktech.com>
Date: Thu, 23 Jul 2020 19:13:17 +0530
Subject: [PATCH 13/13] We have either SPI0 or I2S0, depending on mux

The on-module multiplexer is controlled by pad GPIO_3. #Define a CONFIG
variable, which selects the desired device tree nodes. The (Freescale-
specific) option pinctrl-assert-gpios will then make sure the mux is
steered appropriately.

The hardware would be able to use both, just not simultaneously.

In anticipation of more than one codec implement CONFIG var such that != 0
in general enables I2S and the specific value selects a certain codec. Use
'2' for the LEC-Base R2 codec.

Signed-off-by: Antony Abee Prakash XV <antonyabee.prakashxv@adlinktech.com>
---
 arch/arm/boot/dts/lec-imx6.dtsi | 160 +++++++++++++++++++-------------
 1 file changed, 96 insertions(+), 64 deletions(-)

diff --git a/arch/arm/boot/dts/lec-imx6.dtsi b/arch/arm/boot/dts/lec-imx6.dtsi
index 134ab755e917..6dbffc32b318 100644
--- a/arch/arm/boot/dts/lec-imx6.dtsi
+++ b/arch/arm/boot/dts/lec-imx6.dtsi
@@ -3,6 +3,10 @@
 // Copyright 2012 Freescale Semiconductor, Inc.
 // Copyright 2011 Linaro Ltd.
 
+/* On-module multiplexer: 0 enables SMARC SPI0, !=0 enables I2S0 instead:
+ * 1=TI TLV320AIC23 on LEC-Base R1 carrier, 2=Freescale SGTL5000 on LEC-Base R2 */
+#define CONFIG_I2S_AUDIO       1
+
 #include <dt-bindings/clock/imx6qdl-clock.h>
 #include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/input/input.h>
@@ -101,6 +105,29 @@
 		enable-active-high;
 	};
 
+#if CONFIG_I2S_AUDIO == 2
+	/* Power supplies on the LEC-Base R2 carrier. They cannot be
+	 * controlled, but the SGTL5000 driver won't work without
+	 * regulator definitions. */
+	reg_base_1v8: regulator-base-1v8 {
+		compatible = "regulator-fixed";
+		regulator-name = "base_1v8";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		regulator-boot-on;
+		regulator-always-on;
+	};
+
+	reg_base_3v3: regulator-base-3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "base_3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-boot-on;
+		regulator-always-on;
+	};
+#endif
+
     reg_hdmi: regulator-hdmi {
 		compatible = "regulator-fixed";
 		regulator-name = "hdmi-5v-supply";
@@ -144,32 +171,32 @@
 		};
 	};
 
-	sound {
-		compatible = "fsl,imx6q-sabresd-wm8962",
-			   "fsl,imx-audio-wm8962";
-		model = "wm8962-audio";
-		audio-cpu = <&ssi2>;
+#if CONFIG_I2S_AUDIO == 1
+	sound-i2s {     /* Audio codec on LEC-Base R1 carrier */
+		compatible = "fsl,imx-audio-tlv320aic23";
+		model = "imx-tlv320aic23";
+		ssi-controller = <&ssi2>;
+		mux-int-port = <2>;
+		mux-ext-port = <3>;
 		audio-codec = <&codec>;
-		asrc-controller = <&asrc>;
-		audio-routing =
-			"Headphone Jack", "HPOUTL",
-			"Headphone Jack", "HPOUTR",
-			"Ext Spk", "SPKOUTL",
-			"Ext Spk", "SPKOUTR",
-			"AMIC", "MICBIAS",
-			"IN3R", "AMIC",
-			"DMIC", "MICBIAS",
-			"DMICDAT", "DMIC",
-			"CPU-Playback", "ASRC-Playback",
-			"Playback", "CPU-Playback",
-			"ASRC-Capture", "CPU-Capture",
-			"CPU-Capture", "Capture";
+	};
+#elif CONFIG_I2S_AUDIO == 2
+	sound-i2s {     /* Audio codec on LEC-Base R2 carrier */
+		compatible = "fsl,imx-audio-sgtl5000";
+		model = "imx-sgtl5000";
+		ssi-controller = <&ssi2>;
 		mux-int-port = <2>;
 		mux-ext-port = <3>;
-		codec-master;
-		hp-det-gpios = <&gpio7 8 1>;
-		mic-det-gpios = <&gpio1 9 1>;
+		audio-codec = <&codec>;
+		asrc-controller = <&asrc>;
+		audio-routing =
+			"MIC_IN", "Mic Jack",
+			"Mic Jack", "Mic Bias",
+			"LINE_IN", "Line In Jack",
+			//      "Headphone Jack", "HP_OUT",
+			"Line Out Jack", "LINE_OUT";
 	};
+#endif
 
 	sound-hdmi {
 		compatible = "fsl,imx6q-audio-hdmi",
@@ -322,11 +349,14 @@
         status = "okay";
 };
 
-&audmux {
+#if CONFIG_I2S_AUDIO
+&audmux {	/* SMARC I2S0 (instead of SPI0) */
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_audmux>;
+	pinctrl-assert-gpios = <&gpio1 3 GPIO_ACTIVE_HIGH>; /* AEN# */
 	status = "okay";
 };
+#endif
 
 &clks {
 	assigned-clocks = <&clks IMX6QDL_CLK_LDB_DI0_SEL>,
@@ -347,20 +377,17 @@
 	status = "okay";
 };
 
-&ecspi1 {
-	cs-gpios = <&gpio4 9 0>;
+#if !CONFIG_I2S_AUDIO
+&ecspi1 {	/* SMARC SPI0 (instead of I2S0) */
+	fsl,spi-num-chipselects = <2>;
+	cs-gpios = <&gpio5 25 GPIO_ACTIVE_LOW>, /* SMARC CS0 */
+		   <&gpio3 24 GPIO_ACTIVE_LOW>; /* SMARC CS1 */
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_ecspi1>;
+	pinctrl-assert-gpios = <&gpio1 3 GPIO_ACTIVE_LOW>; /* AEN# */
 	status = "okay";
-
-	flash: m25p80@0 {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		compatible = "st,m25p32", "jedec,spi-nor";
-		spi-max-frequency = <20000000>;
-		reg = <0>;
-	};
 };
+#endif
 
 &fec {
 	pinctrl-names = "default";
@@ -442,28 +469,6 @@
 	sda-gpios = <&gpio5 26 GPIO_ACTIVE_HIGH>;
 	status = "okay";
 
-	codec: wm8962@1a {
-		compatible = "wlf,wm8962";
-		reg = <0x1a>;
-		clocks = <&clks IMX6QDL_CLK_CKO>;
-		DCVDD-supply = <&reg_audio>;
-		DBVDD-supply = <&reg_audio>;
-		AVDD-supply = <&reg_audio>;
-		CPVDD-supply = <&reg_audio>;
-		MICVDD-supply = <&reg_audio>;
-		PLLVDD-supply = <&reg_audio>;
-		SPKVDD1-supply = <&reg_audio>;
-		SPKVDD2-supply = <&reg_audio>;
-		gpio-cfg = <
-			0x0000 /* 0:Default */
-			0x0000 /* 1:Default */
-			0x0013 /* 2:FN_DMICCLK */
-			0x0000 /* 3:Default */
-			0x8014 /* 4:FN_DMICCDAT */
-			0x0000 /* 5:Default */
-		>;
-	};
-
 	accelerometer@1c {
 		compatible = "fsl,mma8451";
 		reg = <0x1c>;
@@ -686,6 +691,29 @@
 			};
 		};
 	};
+
+#if CONFIG_I2S_AUDIO == 1
+	codec: tlv320aic23@1a {         /* Audio codec on LEC-Base R1 carrier */
+		compatible = "ti,tlv320aic23";
+		reg = <0x1a>;
+		clocks = <&clks IMX6QDL_CLK_CKO>; /* SMARC AUDIO_MCK, actually unused on carrier */
+		status = "okay";
+	};
+#endif
+
+#if CONFIG_I2S_AUDIO == 2
+	codec: sgtl5000@0a {            /* Audio codec on LEC-Base R2 carrier */
+		compatible = "fsl,sgtl5000";
+		reg = <0x0a>;
+		clocks = <&clks IMX6QDL_CLK_CKO>; /* SMARC AUDIO_MCK, actually unused on carrier */
+		VDDA-supply = <&reg_base_3v3>;
+		VDDIO-supply = <&reg_base_1v8>;
+		VDDD-supply = <&reg_base_1v8>;
+		micbias-resistor-k-ohms = <4>;
+		micbias-voltage-m-volts = <1250>;
+	};
+#endif
+
 };
 
 &i2c3 {
@@ -751,7 +779,8 @@
 			>;
 		};
 
-		pinctrl_audmux: audmuxgrp {
+#if CONFIG_I2S_AUDIO
+		pinctrl_audmux: audmuxgrp {	/* SMARC I2S0 (instead of SPI0) */
 			fsl,pins = <
 				MX6QDL_PAD_CSI0_DAT7__AUD3_RXD		0x130b0
 				MX6QDL_PAD_CSI0_DAT4__AUD3_TXC		0x130b0
@@ -759,16 +788,17 @@
 				MX6QDL_PAD_CSI0_DAT6__AUD3_TXFS		0x130b0
 			>;
 		};
-
+#else
 		pinctrl_ecspi1: ecspi1grp {
-			fsl,pins = <
-				MX6QDL_PAD_KEY_COL1__ECSPI1_MISO	0x100b1
-				MX6QDL_PAD_KEY_ROW0__ECSPI1_MOSI	0x100b1
-				MX6QDL_PAD_KEY_COL0__ECSPI1_SCLK	0x100b1
-				MX6QDL_PAD_KEY_ROW1__GPIO4_IO09		0x1b0b0
+			fsl,pins = <    /* SMARC SPI0 (instead of I2S0) */
+				MX6QDL_PAD_CSI0_DAT6__ECSPI1_MISO       0x100b1
+				MX6QDL_PAD_CSI0_DAT5__ECSPI1_MOSI       0x100b1
+				MX6QDL_PAD_CSI0_DAT4__ECSPI1_SCLK       0x100b1
+				MX6QDL_PAD_CSI0_DAT7__GPIO5_IO25        0x1b0b0 /* SMARC CS0 */
+				MX6QDL_PAD_EIM_D24__GPIO3_IO24          0x1b0b0 /* SMARC CS1 */
 			>;
 		};
-
+#endif
 		pinctrl_enet: enetgrp {
 			fsl,pins = <
 				MX6QDL_PAD_ENET_MDIO__ENET_MDIO		0x1b0b0
@@ -1180,9 +1210,11 @@
 	status = "okay";
 };
 
-&ssi2 {
+#if CONFIG_I2S_AUDIO
+&ssi2 {	/* SMARC I2S0 (instead of SPI0) */
 	status = "okay";
 };
+#endif
 
 &uart1 {	/* SMARC SER0 */
 	pinctrl-names = "default";
-- 
2.17.1

